modules/siwe.ts

```diff
@@ -23,93 +24,157 @@ if (typeof window !== "undefined" && window.ethereum) {
+  
     /* Create if wallet does not exist in database */
-    const data = {
-      _id: walletAddress
-    };
-    const newEthAccountXhr = await privateBaseAxios.post('/account/ethereum/create', data);
-    console.log({ newEthAccountXhr })
-    const newEthAccount = newEthAccountXhr.data
+    const { data: newEthAccount } = await privateBaseAxios.post('/account/ethereum/create', { _id: walletAddress });
 
     /* Update user_id of newEthAccount */
-    const updateData = {
-      user_id: userId
-    }
-    console.log({ updateData })
-    const updatedAccountXhr = await privateBaseAxios.patch(`/account/ethereum/update/${newEthAccount._id}`, updateData)
-    return updatedAccountXhr.data
+    const { data: updatedAccount } = await privateBaseAxios.patch(
+      `/account/ethereum/update/${newEthAccount._id}`,
+      { user_id }
+    )
+    return updatedAccount
   }
 
-  static async signInWithEthereum(discordId: string) {
-
-    console.log({ signer, provider })
-
-    if (!signer) { alert('Wallet not connected!'); return }
-
-    const walletAddress = await signer.getAddress();
-
-    console.log({ walletAddress })
-
-    /* Get nonce from server */
-    const nonceRes = await privateBaseAxios.get(`/siwe/nonce/${walletAddress}`)
-
-    const nonce = nonceRes.data
-
-    console.log({ nonce })
-
-    const message = this.createSiweMessage(
-      walletAddress,
-      `Sign in with Ethereum to the app. - link to discord_id ${discordId}`,
-      nonce
-    );
-
-    console.log({ message })
+  /**
+   * creates a message and signature and sends for verification.
+   * @param {string} discordId - the discord id
+   * @param {EIP1193Provider} provider optional provider, if not provided, will use provider from window.ethereum
+   * @param {string} address - optional address, if not provided, will use signer
+   * @todo - add error handling and type
+   * @returns {Promise<any>}
+   */
+  static signInWithEthereum(
+    discordId: string,
+    provider?: EIP1193Provider,
+    address?: string
+  ): Promise<any> {
+    return new Promise(async (resolve, reject) => {
+      try {
+        if (!signer) reject("Wallet not connected")
+
+        const walletAddress = address ? ethers.utils.getAddress(address) : await signer.getAddress();
+    
+        /* Get nonce from server */
+        const { data: nonce } = await privateBaseAxios.get(`/siwe/nonce/${walletAddress}`)
+    
+        /* Create message */
+        const message = this.createSiweMessage(
+          walletAddress,
+          `Sign in with Ethereum to the app. - link to discord_id ${discordId}`,
+          nonce
+        );
+
+        // generate signature. If address is provided, use provider, otherwise use signer
+        const signature = await this.generateSignature(message, provider, address);
+    
+        resolve(await this.sendForVerification(message, signature, discordId, walletAddress));
+      } catch (e) {
+        reject(e);
+      }
+    });
+  }
```

Comment:
Switch this to return a Promise explicitly, so I can reject and pass errors up to the client method.